pkg/storegateway/indexcache/remote.go

```diff
@@ -136,11 +147,65 @@ func (c *RemoteIndexCache) FetchMultiPostings(ctx context.Context, userID string
 	return hits, misses
 }
 
+// postingsCacheKey returns the cache key used to store postings matching the input
+// label name/value pair in the given block.
 func postingsCacheKey(userID string, blockID ulid.ULID, l labels.Label) string {
+	const (
+		prefix    = "P2:"
+		separator = ":"
+	)
+
+	// Compute the label hash.
+	lblHash := postingsCacheKeyLabelHash(l)
+
+	// Preallocate the byte slice used to store the cache key.
+	expectedLen := len(prefix) + len(userID) + 1 + ulid.EncodedSize + 1 + base64.RawStdEncoding.EncodedLen(blake2b.Size256)
+	key := make([]byte, expectedLen)
+	offset := 0
+
+	offset += copy(key[offset:], prefix)
```

Comment:
nit: this seems very similar to what `bytes.Buffer` does

Reply:
That's true, but I haven't found an easy way to do `base64.RawURLEncoding.Encode()` without an extra allocation. Also `bytes.Buffer.String()` does an extra allocation too.